proxy.go

```diff
@@ -98,246 +94,150 @@ func (p *Proxy) setupProxy() {
 	// See: https://github.com/golang/go/issues/28866
 	// See: https://github.com/elazarl/goproxy/issues/306
 	p.proxy.KeepHeader = true
-	p.proxy.Tr = &http.Transport{}
 
-	// Local proxy authentication.
-	if u := p.config.LocalProxyURI.User; u.Username() != "" {
-		p.setupBasicAuth(u)
-	}
-
-	p.setupProxyHandlers()
-}
+	p.setupLocalhostProxy()
 
-// Config returns a copy of the proxy configuration.
-func (p *Proxy) Config() *ProxyConfig {
-	return p.config.Clone()
-}
-
-// Mode returns mode of operation of the proxy as specified in the config.
-func (p *Proxy) Mode() Mode {
-	switch {
-	case p.config.UpstreamProxyURI != nil:
-		return Upstream
-	case p.config.PACURI != nil:
-		return PAC
+	switch p.Mode() {
+	case Direct:
+		p.setupDirect()
+	case Upstream:
+		p.setupUpstreamProxy()
+	case PAC:
+		p.setupPACProxy()
 	default:
-		return Direct
+		panic(fmt.Errorf("unknown mode %q", p.Mode()))
 	}
-}
 
-// Sets the `Proxy-Authorization` header based on `uri` user info.
-func (p *Proxy) setProxyBasicAuthHeader(uri *url.URL, req *http.Request) {
-	req.Header.Set(
-		proxyAuthHeader,
-		fmt.Sprintf("Basic %s", userInfoBase64(uri.User)),
-	)
-
-	p.log.Debugf(
-		"%s header set with %s:*** for %s",
-		proxyAuthHeader,
-		uri.User.Username(),
-		req.URL.String(),
-	)
+	p.setupProxyBasicAuth()
+	p.setupSiteBasicAuth()
 }
 
-// Removes any upstream proxy settings.
-func resetUpstreamSettings(ctx *goproxy.ProxyCtx) {
-	ctx.Proxy.ConnectDial = nil
+func (p *Proxy) setupLocalhostProxy() {
```

Comment:
I think this changed the logic for `localhost` handling. It's possible the original code was wrong as well. :)

The intention is to only pass requests for `localhost` to a parent proxy (set via upstream or PAC) if specifically enabled. 

However, if *disabled*, requests coming in to `localhost` should be sent to localhost directly regardless of the parent proxy settings. If I'm reading this correctly, if `p.config.ProxyLocalhost == false` and a request came in, it would return a 502 error.

If there is no parent proxy set, it would always go to `localhost` regardless of `p.config.ProxyLocalhost`.

Reply:
Precisely. Let's take a look at it case by case:

OLD

|Mode|ProxyLocalhost=true|ProxyLocalhost=false|
|--|--|--|
|DIRECT       | Dial localhost  | Dial localhost |
|UPSTREAM| Proxy localhost | Dial localhost |

NEW

|Mode|ProxyLocalhost=true|ProxyLocalhost=false|
|--|--|--|
|DIRECT       | Dial localhost  | Bad GW |
|UPSTREAM| Proxy localhost | Bad GW |

The original behavior has two problems
1. It's inconsistent
1. It is security threat, you think the localhost is not exposed but it is